core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -227,6 +237,39 @@ public void close() {
     }
   }
 
+  private static class PositionStreamDeleteMarker<T> extends PositionStreamDeleteFilter<T> {
+    private final Consumer<T> markDeleted;
+
+    private PositionStreamDeleteMarker(CloseableIterable<T> rows, Function<T, Long> extractPos,
+                                       CloseableIterable<Long> deletePositions, Consumer<T> markDeleted) {
+      super(rows, extractPos, deletePositions);
+      this.markDeleted = markDeleted;
+    }
+
+    @Override
+    protected PositionFilterIterator createPosDeleteIterator(CloseableIterator<T> items,
+                                                             CloseableIterator<Long> deletePosIterator) {
+      return new PositionDeleteMarkerIterator(items, deletePosIterator);
+    }
+
+    private class PositionDeleteMarkerIterator extends PositionFilterIterator {
```

Comment:
I probably need to read fully , but I also feel on the same line that we can have the caller use functional composition, instead of trying to have one master iterator that does everything?  ie,

``` filter.filter(rows).transform(deleteMarker)```

It would be cleaner and avoid having this complex iterator.

Reply:
Yeah, as I said, we couldn't combine filter and deleteMarker. The current way is the simplest way I tried. Maybe there is a better way to handle this. I'm open to ideas.

Reply:
One idea is composition:

Make a class DeleteIterator(Iterator<T> items, Iterator<T> deletePosIterator), which has method isDeleted() (the inverse of today's PositionFilterIterator's shouldKeep).

Then both PosFilterIterator and PosDeleteIterator compose it:
```
PosFilterIterator(DeleteIterator<T> deleteIterator) extend FilterIterator {
   boolean shouldKeep(row) {
     return !deleteIterator.isDeleted(row) 
   }

PositionDeleteMarkerIterator(DeleteIterator<T> deleteIterator, Consumer<T> markDeleted) {
   T next() {
      if (deleteIterator.isDeleted(row)) {
         markDeleted.apply(row)
      }
      return deleteIterator.next()
   }
```

Hope that's correct.  I think overhead wise, it should be similar today to the extra super method call?  Also fyi if @aokolnychyi  has any thought on it.

Reply:
Thanks for the suggestion. Made the change. We don't have to create the new class for the marker. I use the `CloseableIterator.transform` for the marker, and an anonymous class of FilterIterator for the filter. Added a test as well.